File: swift/hpke.swift
/*
 * Copyright (c) 2012-2020 MIRACL UK Ltd.
 *
 * This file is part of MIRACL Core
 * (see https://github.com/miracl/core).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//
//  hpke.swift
//
//  Created by Michael Scott on 19/12/2019.
//  Copyright (c) 2019 Michael Scott. All rights reserved.
//

import core

/* HPKE high-level API functions  */
/*
public func printBinary(_ array: [UInt8])
{
    for i in 0 ..< array.count
    {
        let h=String(format:"%02x",array[i])
        print("\(h)", terminator: "")
    }
    print(" ")
}
*/
public struct HPKE
{

    static func LabeledExtract(_ salt:[UInt8]?,_ suite_ID:[UInt8],_ label:String,_ ikm:[UInt8]?) -> [UInt8] {
        let rfc="HPKE-v1"
        let RFC=[UInt8](rfc.utf8)
        let LABEL=[UInt8](label.utf8)
        var likm=[UInt8]();
        for i in 0..<RFC.count {
            likm.append(RFC[i])
        }
        for i in 0..<suite_ID.count {
            likm.append(suite_ID[i])
        }
        for i in 0..<LABEL.count {
            likm.append(LABEL[i])
        }
        if ikm != nil {
            for i in 0..<ikm!.count {
                likm.append(ikm![i])
            }
        }
        return HMAC.HKDF_Extract(HMAC.MC_SHA2,CONFIG_CURVE.HASH_TYPE,salt,likm)
    }

    static func LabeledExpand(_ prk:[UInt8],_ suite_ID:[UInt8],_ label:String,_ info:[UInt8]?,_ L:Int) -> [UInt8] {
        let rfc="HPKE-v1"
        let RFC=[UInt8](rfc.utf8)
        let LABEL=[UInt8](label.utf8)
        let ar = HMAC.inttoBytes(L, 2)
        var linfo=[UInt8]();
        for i in 0..<ar.count {
            linfo.append(ar[i])
        }
        for i in 0..<RFC.count {
            linfo.append(RFC[i])
        }
        for i in 0..<suite_ID.count {
            linfo.append(suite_ID[i])
        }
        for i in 0..<LABEL.count {
            linfo.append(LABEL[i])
        }
        if info != nil {
            for i in 0..<info!.count {
                linfo.append(info![i])
            }
        }
        return HMAC.HKDF_Expand(HMAC.MC_SHA2,CONFIG_CURVE.HASH_TYPE,L,prk,linfo)
    }

    static func ExtractAndExpand(_ config_id: Int,_ dh:[UInt8],_ context:[UInt8]) -> [UInt8] {
        let kem = config_id&255
        let txt="KEM"        
        var suite_ID=[UInt8]();
        let kem_id=HMAC.inttoBytes(kem, 2)
        let ckem=[UInt8](txt.utf8)
        for i in 0..<ckem.count {
            suite_ID.append(ckem[i])
        }
        suite_ID.append(kem_id[0])
        suite_ID.append(kem_id[1])

        let prk=LabeledExtract(nil,suite_ID,"eae_prk",dh)
        return LabeledExpand(prk,suite_ID,"shared_secret",context,CONFIG_CURVE.HASH_TYPE)
    }

    @discardableResult static public func DeriveKeyPair(_ config_id: Int,_ sk:inout [UInt8],_ pk:inout [UInt8],_ seed:[UInt8]) -> Bool {
        var counter=0
        let kem = config_id&255
        var NULLRNG : RAND? = nil

        let txt="KEM"        
        var suite_ID=[UInt8]();
        let kem_id=HMAC.inttoBytes(kem, 2)
        let ckem=[UInt8](txt.utf8)
        for i in 0..<ckem.count {
            suite_ID.append(ckem[i])
        }
        suite_ID.append(kem_id[0])
        suite_ID.append(kem_id[1])

        let prk=LabeledExtract(nil,suite_ID,"dkp_prk",seed)
        var s=[UInt8]();
        if kem==32 || kem==33 { // RFC7748
            s=LabeledExpand(prk,suite_ID,"sk",nil,ECDH.EGS);
            s.reverse();
		    if kem==32 {
			    s[ECDH.EGS-1]&=248
			    s[0]&=127
			    s[0]|=64
		    } else {
			    s[ECDH.EGS-1]&=252
			    s[0]|=128
		    }
        } else {
            var bit_mask=0xff;
            if kem==18 {
                bit_mask=1;
            }
            for _ in 0..<ECDH.EGS {
                s.append(0);
            }
            while !ECDH.IN_RANGE(s) && counter<256 {
                var info=[UInt8](repeating: 0, count:1)
                info[0]=UInt8(counter)
                s=LabeledExpand(prk,suite_ID,"candidate",info,ECDH.EGS)
                s[0] &= UInt8(bit_mask)
                counter = counter+1
            }
        }
        for i in 0..<ECDH.EGS {
            sk[i]=s[i]
        }
        ECDH.KEY_PAIR_GENERATE(&NULLRNG, &sk, &pk)
        if kem==32 || kem==33 {
            pk.reverse()
        }
        if counter<256 {
            return true
        }
        return false
    }

        var T=[UInt8](repeating: 0,count: ECDH.EFS)

    static public func ENCAP(_ config_id: Int,_ skE:[UInt8],_ pkE:[UInt8],_ pkR:[UInt8]) -> [UInt8] {
        var dh=[UInt8](repeating: 0,count: ECDH.EFS)
        var kemcontext=[UInt8]()

        let kem = config_id&255
  
        if kem==32 || kem==33 {
            ECDH.ECPSVDP_DH(skE, pkR.reversed(), &dh, 0)
            dh.reverse()
        } else {
            ECDH.ECPSVDP_DH(skE, pkR, &dh, 0)
        }

        kemcontext.append(contentsOf: pkE)
        kemcontext.append(contentsOf: pkR)

        return ExtractAndExpand(config_id,dh,kemcontext)
    }

    static public func DECAP(_ config_id: Int,_ skR: [UInt8],_ pkE:[UInt8],_ pkR:[UInt8]) -> [UInt8] {
        var dh=[UInt8](repeating: 0,count: ECDH.EFS)
        var kemcontext=[UInt8]()

        let kem = config_id&255

        if kem==32 || kem==33 {
            ECDH.ECPSVDP_DH(skR, pkE.reversed(), &dh, 0)
            dh.reverse()
        } else {
            ECDH.ECPSVDP_DH(skR, pkE, &dh, 0)
        }
 
        kemcontext.append(contentsOf: pkE)
        kemcontext.append(contentsOf: pkR)

        return ExtractAndExpand(config_id,dh,kemcontext)
    }

    static public func AUTHENCAP(_ config_id: Int,_ skE:[UInt8],_ skS:[UInt8],_ pkE:[UInt8],_ pkR:[UInt8],_ pkS:[UInt8]) -> [UInt8] {
        var dh=[UInt8](repeating: 0,count: ECDH.EFS)
        var dh1=[UInt8](repeating: 0,count: ECDH.EFS)
        var kemcontext=[UInt8]()
        let kem = config_id&255

        if kem==32 || kem==33 {
            ECDH.ECPSVDP_DH(skE, pkR.reversed(), &dh, 0)
            ECDH.ECPSVDP_DH(skS, pkR.reversed(), &dh1, 0)
            dh.reverse()
            dh1.reverse()
        } else {
            ECDH.ECPSVDP_DH(skE, pkR, &dh, 0)
            ECDH.ECPSVDP_DH(skS, pkR, &dh1, 0)
        }

        dh.append(contentsOf: dh1)

        kemcontext.append(contentsOf: pkE)
        kemcontext.append(contentsOf: pkR)
        kemcontext.append(contentsOf: pkS)
        return ExtractAndExpand(config_id,dh,kemcontext)
    }

    static public func AUTHDECAP(_ config_id: Int,_ skR:[UInt8],_ pkE: [UInt8],_ pkR:[UInt8],_ pkS: [UInt8]) -> [UInt8]  {
        var dh=[UInt8](repeating: 0,count: ECDH.EFS)
        var dh1=[UInt8](repeating: 0,count: ECDH.EFS)
        var kemcontext=[UInt8]()

        let kem = config_id&255

        if kem==32 || kem==33 {
            ECDH.ECPSVDP_DH(skR, pkE.reversed(), &dh, 0)
            ECDH.ECPSVDP_DH(skR, pkS.reversed(), &dh1, 0)
            dh.reverse() 
            dh1.reverse()
        } else {
            ECDH.ECPSVDP_DH(skR, pkE, &dh, 0)
            ECDH.ECPSVDP_DH(skR, pkS, &dh1, 0)
        }

        dh.append(contentsOf: dh1)

        kemcontext.append(contentsOf: pkE)
        kemcontext.append(contentsOf: pkR)
        kemcontext.append(contentsOf: pkS)
        return ExtractAndExpand(config_id,dh,kemcontext)
    }

    static public func KEYSCHEDULE(_ config_id: Int,_ mode: Int,_ Z: [UInt8],_ info: [UInt8],_ psk: [UInt8]?,_ pskID: [UInt8]?) -> ([UInt8],[UInt8],[UInt8]) {
        var context=[UInt8]()

        let kem = config_id&255
        let kdf=(config_id>>8)&3
        let aead=(config_id>>10)&3
        let txt="HPKE"        
        var suite_ID=[UInt8]();
        let ckem=[UInt8](txt.utf8)
        for i in 0..<ckem.count {
            suite_ID.append(ckem[i])
        }
        var num=HMAC.inttoBytes(kem, 2)
        suite_ID.append(num[0])
        suite_ID.append(num[1])
        num=HMAC.inttoBytes(kdf, 2)
        suite_ID.append(num[0])
        suite_ID.append(num[1])
        num=HMAC.inttoBytes(aead, 2)
        suite_ID.append(num[0])
        suite_ID.append(num[1])

        let ar = HMAC.inttoBytes(mode, 1)
        context.append(contentsOf: ar)

        var H=LabeledExtract(nil,suite_ID,"psk_id_hash",pskID!)
        context.append(contentsOf: H)
        H=LabeledExtract(nil,suite_ID,"info_hash",info)
        context.append(contentsOf: H)
        //H=LabeledExtract(nil,suite_ID,"psk_hash",psk!);
        
        //let secret=LabeledExtract(H,suite_ID,"secret",Z)
        let secret=LabeledExtract(Z,suite_ID,"secret",psk!);

        let key=LabeledExpand(secret,suite_ID,"key",context,CONFIG_CURVE.AESKEY)
        let nonce=LabeledExpand(secret,suite_ID,"base_nonce",context,12)
        let exp_secret=LabeledExpand(secret,suite_ID,"exp",context,CONFIG_CURVE.HASH_TYPE)

        return (key,nonce,exp_secret)
    }
}

File: java/HPKE.java
/*
 * Copyright (c) 2012-2020 MIRACL UK Ltd.
 *
 * This file is part of MIRACL Core
 * (see https://github.com/miracl/core).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* Hybrid Public Key Encryption */

/* Following https://datatracker.ietf.org/doc/draft-irtf-cfrg-hpke/?include_text=1 */

package org.miracl.core.XXX;

import org.miracl.core.RAND;
import org.miracl.core.HMAC;
import org.miracl.core.GCM;

public final class HPKE {

    static final int GROUP = ECDH.EGS;

/*
    private static void printBinary(byte[] array) {
        int i;
        for (i = 0; i < array.length; i++) {
            System.out.printf("%02x", array[i]);
        }
        System.out.println();
    }
*/
    static void reverse(byte[] X) {
	    int lx=X.length;
	    for (int i=0;i<lx/2;i++)  {
		    byte ch =X[i];
		    X[i]=X[lx-i-1];
		    X[lx-i-1]=ch;
	    }
    }

    static byte[] LabeledExtract(byte[] SALT,byte[] SUITE_ID,String label,byte[] IKM) {
        String rfc="HPKE-v1";

        byte[] RFC=rfc.getBytes();
        byte[] LABEL=label.getBytes();
        int IKMlen;
        if (IKM==null)
            IKMlen=0;
        else
            IKMlen=IKM.length;
        byte[] LIKM = new byte[RFC.length+SUITE_ID.length+LABEL.length+IKMlen];
        int k=0;
        for (int i=0;i<RFC.length;i++)
            LIKM[k++]=RFC[i];
        for (int i=0;i<SUITE_ID.length;i++)
            LIKM[k++]=SUITE_ID[i];
        for (int i=0;i<LABEL.length;i++)
            LIKM[k++]=LABEL[i];
        if (IKM!=null) {
            for (int i=0;i<IKM.length;i++)
                LIKM[k++]=IKM[i];
        }

        return HMAC.HKDF_Extract(HMAC.MC_SHA2,CONFIG_CURVE.HASH_TYPE,SALT,LIKM);
    }

    static byte[] LabeledExpand(byte[] PRK,byte[] SUITE_ID,String label,byte[] INFO,int L) {
        byte[] AR = HMAC.inttoBytes(L, 2);
        String rfc="HPKE-v1";
        byte[] RFC=rfc.getBytes();
        byte[] LABEL=label.getBytes();
        int INFOlen;
        if (INFO==null)
            INFOlen=0;
        else
            INFOlen=INFO.length;
        byte[] LINFO=new byte[2+RFC.length+SUITE_ID.length+LABEL.length+INFOlen];
        LINFO[0]=AR[0];
        LINFO[1]=AR[1];
        int k=2;
        for (int i=0;i<RFC.length;i++)
            LINFO[k++]=RFC[i];
        for (int i=0;i<SUITE_ID.length;i++)
            LINFO[k++]=SUITE_ID[i];
        for (int i=0;i<LABEL.length;i++)
            LINFO[k++]=LABEL[i];
        if (INFO!=null) {
            for (int i=0;i<INFO.length;i++)
                LINFO[k++]=INFO[i];
        }

        return HMAC.HKDF_Expand(HMAC.MC_SHA2,CONFIG_CURVE.HASH_TYPE,L,PRK,LINFO);
    }

    static byte[] ExtractAndExpand(int config_id,byte[] DH,byte[] CONTEXT) {
	    int kem = config_id&255;
        String txt="KEM";
        byte[] KEM_ID = HMAC.inttoBytes(kem, 2);
        byte[] KEM =txt.getBytes();
        byte[] SUITE_ID = new byte[2+KEM.length];
        int k=KEM.length;
        for (int i=0;i<k;i++)
            SUITE_ID[i]=KEM[i];
        SUITE_ID[k]=KEM_ID[0];
        SUITE_ID[k+1]=KEM_ID[1];

        byte[] PRK=LabeledExtract(null,SUITE_ID,"eae_prk",DH);
        return LabeledExpand(PRK,SUITE_ID,"shared_secret",CONTEXT,CONFIG_CURVE.HASH_TYPE);
    }

    public static boolean DeriveKeyPair(int config_id,byte[] SK,byte[] PK,byte[] SEED) {
        int counter=0;
        byte[] INFO=new byte[1];
	    int kem = config_id&255;
        String txt="KEM";
        byte[] KEM_ID = HMAC.inttoBytes(kem, 2);
        byte[] KEM =txt.getBytes();
        byte[] SUITE_ID = new byte[2+KEM.length];
        int k=KEM.length;
        for (int i=0;i<k;i++)
            SUITE_ID[i]=KEM[i];
        SUITE_ID[k]=KEM_ID[0];
        SUITE_ID[k+1]=KEM_ID[1];
        byte[] PRK=LabeledExtract(null,SUITE_ID,"dkp_prk",SEED);
        byte[] S;
		if (kem==32 || kem==33) { // RFC7748
           S=LabeledExpand(PRK,SUITE_ID,"sk",null,GROUP);
           reverse(S);
           if (kem==32)
           {
	            S[GROUP-1]&=248;
				S[0]&=127;
				S[0]|=64;
           } else {
				S[GROUP-1]&=252;
				S[0]|=128;
           }
        } else {
            int bit_mask;
            if (kem==18) bit_mask=1;
            else bit_mask=0xff;
            S=new byte[GROUP];
            while (!ECDH.IN_RANGE(S) && counter<256)
            {
                INFO[0]=(byte)counter;
                S=LabeledExpand(PRK,SUITE_ID,"candidate",INFO,GROUP);
                S[0]&=bit_mask;
                counter++;
            }
        }
        for (int i=0;i<GROUP;i++)
            SK[i]=S[i];
	    ECDH.KEY_PAIR_GENERATE(null, SK, PK);  
        if (kem==32 || kem==33)
            reverse(PK);
        if (counter<256) return true;
        return false;
    }

    public static byte[] encap(int config_id,byte[] skE,byte[] pkE,byte[] pkR) {
        byte[] DH = new byte[ECDH.EFS/*pkE.length*/];
        byte[] kemcontext = new byte[2*pkE.length];
	    int kem = config_id&255;

		if (kem==32 || kem==33) {
			reverse(pkR);
	        ECDH.SVDP_DH(skE, pkR, DH, 0);
            reverse(pkR);
            reverse(DH);
		} else {
	        ECDH.SVDP_DH(skE, pkR, DH, 0);
        }
	 
        int k=0;
        for (int i=0;i<pkE.length;i++)
            kemcontext[k++]=pkE[i];
        for (int i=0;i<pkR.length;i++)
            kemcontext[k++]=pkR[i];

    	return ExtractAndExpand(config_id,DH,kemcontext);
    }

    public static byte[] decap(int config_id,byte[] skR,byte[] pkE,byte[] pkR) {
        byte[] DH = new byte[ECDH.EFS/*pkE.length*/];
        byte[] kemcontext = new byte[2*pkE.length];
	    int kem = config_id&255;

	    if (kem==32 || kem==33) {
		    reverse(pkE);
	        ECDH.SVDP_DH(skR, pkE, DH, 0);
            reverse(pkE);
            reverse(DH);
	    } else {
            ECDH.SVDP_DH(skR, pkE, DH, 0);
        }

        int k=0;
        for (int i=0;i<pkE.length;i++)
            kemcontext[k++]=pkE[i];
        for (int i=0;i<pkR.length;i++)
            kemcontext[k++]=pkR[i];

    	return ExtractAndExpand(config_id,DH,kemcontext);
    }

    public static byte[] authEncap(int config_id,byte[] skE,byte[] skS,byte[] pkE,byte[] pkR,byte[] pkS) {
        int pklen=pkE.length;
        byte[] DH = new byte[ECDH.EFS];
        byte[] DH1 = new byte[ECDH.EFS];
        byte[] kemcontext = new byte[3*pklen];
	
	    int kem = config_id&255;

		if (kem==32 || kem==33) {
			reverse(pkR);
	        ECDH.SVDP_DH(skE, pkR, DH, 0);
	        ECDH.SVDP_DH(skS, pkR, DH1, 0);
			reverse(pkR);
            reverse(DH);
            reverse(DH1);
		} else {
	        ECDH.SVDP_DH(skE, pkR, DH, 0);
	        ECDH.SVDP_DH(skS, pkR, DH1, 0);
        }

	    byte[] ZZ = new byte[2*ECDH.EFS];
	    for (int i=0;i<ECDH.EFS;i++) {
		    ZZ[i] = DH[i];
            ZZ[ECDH.EFS+i]= DH1[i];
	    }

	    for (int i=0;i<pklen;i++) {
		    kemcontext[i] = pkE[i];
            kemcontext[pklen+i]= pkR[i];
            kemcontext[2*pklen+i]= pkS[i];
	    }

	    return ExtractAndExpand(config_id,ZZ,kemcontext);
    }

    public static byte[] authDecap(int config_id,byte[] skR,byte[] pkE,byte[] pkR,byte[] pkS) {
        int pklen=pkE.length;
        byte[] DH = new byte[ECDH.EFS];
        byte[] DH1 = new byte[ECDH.EFS];
        byte[] kemcontext = new byte[3*pklen];

		int kem = config_id&255;

	    if (kem==32 || kem==33) {
		    reverse(pkE);
		    reverse(pkS);
    	    ECDH.SVDP_DH(skR, pkE, DH, 0);
	        ECDH.SVDP_DH(skR, pkS, DH1, 0);
            reverse(pkE);
            reverse(pkS);
            reverse(DH);
            reverse(DH1);
	    } else {
	        ECDH.SVDP_DH(skR, pkE, DH, 0);
	        ECDH.SVDP_DH(skR, pkS, DH1, 0);
        }

	    byte[] ZZ = new byte[2*ECDH.EFS];
	    for (int i=0;i<ECDH.EFS;i++) {
		    ZZ[i] = DH[i];
            ZZ[ECDH.EFS+i]= DH1[i];
	    }

	    for (int i=0;i<pklen;i++) {
		    kemcontext[i] = pkE[i];
            kemcontext[pklen+i]= pkR[i];
            kemcontext[2*pklen+i]= pkS[i];
	    }
	    return ExtractAndExpand(config_id,ZZ,kemcontext);
    }

    public static void keySchedule(int config_id,byte[] key,byte[] nonce,byte[] exp_secret,int mode,byte[] Z,byte[] info,byte[] psk,byte[] pskID) {
        int ctxlen=1+2*CONFIG_CURVE.HASH_TYPE;
	    byte[] context = new byte[ctxlen];
        byte[] num;	

	    int kem = config_id&255;
	    int kdf=(config_id>>8)&3;
	    int aead=(config_id>>10)&3;
        String txt="HPKE"; 
        byte[] KEM =txt.getBytes();
        byte[] SUITE_ID = new byte[6+KEM.length];
        int k=KEM.length;
        for (int i=0;i<k;i++)
            SUITE_ID[i]=KEM[i];
        num = HMAC.inttoBytes(kem, 2);
        SUITE_ID[k++]=num[0];
        SUITE_ID[k++]=num[1];
        num = HMAC.inttoBytes(kdf, 2);
        SUITE_ID[k++]=num[0];
        SUITE_ID[k++]=num[1];
        num = HMAC.inttoBytes(aead, 2);
        SUITE_ID[k++]=num[0];
        SUITE_ID[k++]=num[1];
        k=0;
	    byte[] ar = HMAC.inttoBytes(mode, 1);
        for (int i=0;i<ar.length;i++)
		    context[k++] = ar[i];

        byte[] H=LabeledExtract(null,SUITE_ID,"psk_id_hash",pskID);
        for (int i=0;i<CONFIG_CURVE.HASH_TYPE;i++)
            context[k++]=H[i];
        H=LabeledExtract(null,SUITE_ID,"info_hash",info);
        for (int i=0;i<CONFIG_CURVE.HASH_TYPE;i++)
            context[k++]=H[i];
//System.out.print("context= 0x"); printBinary(context);
//        H=LabeledExtract(null,SUITE_ID,"psk_hash",psk);
//        byte[] secret=LabeledExtract(H,SUITE_ID,"secret",Z);

        byte[] secret=LabeledExtract(Z,SUITE_ID,"secret",psk);

//System.out.print("secret= 0x"); printBinary(secret);
        byte[] ex=LabeledExpand(secret,SUITE_ID,"key",context,CONFIG_CURVE.AESKEY);
        for (int i=0;i<ex.length;i++) key[i]=ex[i];

        ex=LabeledExpand(secret,SUITE_ID,"base_nonce",context,12);
        for (int i=0;i<ex.length;i++) nonce[i]=ex[i];

        if (exp_secret!=null)
        {
            ex=LabeledExpand(secret,SUITE_ID,"exp",context,CONFIG_CURVE.HASH_TYPE);
            for (int i=0;i<ex.length;i++) exp_secret[i]=ex[i];
        }
    }
}


